doc/guides/dts/index.rst

```diff
@@ -346,44 +346,534 @@ Below is sample header content generated for this node, in
-"the first node with compatible string ``nxp,kinetis-sim``.
+Macros generated from particular devicetree nodes start with ``DT_<node>``,
+where ``<node>`` identifies the node. Several identifiers are generated for
+nodes:
 
-Aliases are also generated from any properties in the ``/aliases`` node. Take
-the DTS fragment below as an example.
+``DT_(<bus>_)<compatible>_<unit-address>``
+    The compatible string for the node followed by its unit address. If the
+    node has several compatible strings, the first one that matched a binding
+    is used.
+
+    If the node appears on a bus (has ``on-bus:`` in its binding), then the
+    compatible string and unit address of the bus node is put before the
+    compatible string for the node itself.
+
+    For nodes that don't have a unit address (no ``...@<address>`` in the node
+    name), the unit address of the parent node is used instead. If the parent
+    node has no unit address either, the name of the node is used for
+    ``<unit-address>``, as a fallback.
+
+    For the ``sim@40047000`` node above, the compatible string is
+    ``"nxp,kinetis-sim"``. This gets converted to ``NXP_KINETIS_SIM`` by
+    uppercasing and replacing non-alphanumeric characters with underscores.
+    Adding the unit address gives the node prefix
+    ``DT_NXP_KINETIS_SIM_40047000``.
+
+``DT_INST_<instance-no.>_<compatible>``
+    The compatible string for the node together with an instance number.
+
+    The instance number is a unique index among all enabled
+    (``status = "okay"``) nodes that have a particular compatible string,
+    counting from zero. For example, if there are two enabled nodes that have
+    ``compatible = "foo,uart"``, then these node prefixes get generated:
+
+    .. code-block:: none
+
+       DT_INST_0_FOO_UART
+       DT_INST_1_FOO_UART
+
+``DT_ALIAS_<alias>``
+    Generated from the names of any properties in the ``/aliases`` node that
+    point to the node. See :ref:`dt-alias-chosen` for an overview.
+
+    For example, assume ``/aliases`` looks like this:
+
+    .. code-block:: none
+
+       aliases {
+               uart-1 = &uart1;
+       };
+
+    The alias name (``i2c-1``) is converted to ``I2C_1`` by uppercasing and
+    replacing non-alphanumeric characters with underscores, generating the node
+    prefix ``DT_ALIAS_UART_1`` for the ``&uart1`` node.
+
+    .. note::
+
+       Currently, an older deprecated ``DT_<compatible>_<alias>`` form is also
+       generated for aliases. For the example above, assuming the compatible
+       string for the ``&uart1`` node is ``"foo,uart"``, this gives
+       ``DT_FOO_UART_UART_1``.
+
+       Work is underway to replace this form with ``DT_ALIAS_*``.
+
+Property identifiers in macro names
+-----------------------------------
+
+Macros for particular properties on nodes have the form
+``DT_<node>_<property>``. ``<node>`` is one the node prefixes from above, and
+``<property>`` identifies the property.
```

Comment:
It would be good to have a full example here

Reply:
What are you thinking w/regards to an example here?  The specific `property` types have examples below.

Reply:
Agreed, I was thinking about a generic example but it's covered extensively later